[WIP] `df.apply`: add support for `engine='bodo'`

scott-routledge2 opened the pull request:
Adds support for specifying engine="bodo" when executing UDF's on DataFrames via `df.apply`.

- [x] closes #60668
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Commit:
add basic support for engine=bodo, df.apply

scott-routledge2 commented on pandas/tests/apply/test_bodo.py:
```diff
@@ -0,0 +1,105 @@
+
+    func = "mean"
+    axis = 1
+    result = df.apply(func, axis)
```
```suggestion
    result = df.apply(func, axis, engine="bodo")
```

Commit:
Merge remote-tracking branch 'upstream/main' into scott/bodo_udf_engine

Commit:
fix test

Commit:
adjust minimum version requirements

Commit:
update ci envs

Commit:
add channel

Commit:
try skipping some tests

WillAyd commented:
Is there an issue open to discuss this? Adding another engine internally is a non-trivial amount of work and maintenance, so I'm not sure we would even add this without discussion up front.

scott-routledge2 commented:
> Is there an issue open to discuss this? Adding another engine internally is a non-trivial amount of work and maintenance, so I'm not sure we would even add this without discussion up front.

@WillAyd Hi Will, thank you for the feedback, I would be happy to open an issue and start a conversation. The intention of this draft was mainly experimentation.

datapythonista reviewed (commented):
Changes look good to me. I added some comments with ideas, but just minor things.

It'd still be good to have an issue for this (or maybe just update the description of the PR), so everybody understands in detail why this is being proposed, and what use cases this addresses. But in general, this seems good.

datapythonista committed:
Merge branch 'main' into scott/bodo_udf_engine

Commit:
apply feedback

WillAyd commented on pandas/tests/apply/test_bodo.py:
```diff
@@ -0,0 +1,107 @@
+import numpy as np
```
Rather than creating separate tests for bodo is there a way that we can create a fixture for the three different engines? Structuring the tests that way would be very helpful to ensure result consistency

scott-routledge2 replied:
I agree this would be ideal. Part of the reason I structured it this way was that it made it easier to test bodo engine specific behavior e.g. error messages and to match tests for the numba engine. Though I did notice there was also already an engine fixture in `test_frame_apply.py`. I tried adding bodo to this fixture but found that it changed the behavior numba engine tests due to bodo installing some numba extensions.

WillAyd replied:
Hmm OK thanks for the explanation. So do you think those behavioral changes are a bug with bodo or Numba?

W.r.t. the error messages, it would be nice if we catch engine specific errors and offer our own abstraction on top of this. You can see our existing errors in the pandas.errors module - some of those might be useable for issues here, or we may need to create a `ExecutionError` to abstract engine failures

scott-routledge2 replied:
I don't necessarily think this is a bug with either bodo or numba and the changes are fairly benign AFAICT. Although I'd be happy to dive deeper into them if needed.

For the errors, I was trying to catch cases that are not supported by the current Bodo UDF engine ahead of time and raise `NotImplementedError` which is what Numba does here, although in other places they use `NumbaUtilError` for unsupported numba engine routines. When you say "catch engine specific errors" does that mean catching any exceptions originating from Bodo/Numba internally during compiling/running the jitt'ed function and raise `ExecutionError` with the captured message? 

Thanks!

Commit:
debug failing unit tests and apply style comments

Commit:
Merge remote-tracking branch 'upstream/main' into scott/bodo_udf_engine

Commit:
Fix tests and add docstrings

Commit:
Merge remote-tracking branch 'upstream/main' into scott/bodo_udf_engine

Commit:
run bodo tests in separate pytest session

Commit:
remove workaround in conftest.py

Commit:
remove bodo from env file for window CI

Commit:
remove bodo from min version CI

Commit:
add ExecutionError and raise from BodoError

Commit:
remove bodo from min version CI

Commit:
fix typo in min version string

scott-routledge2 commented:
@datapythonista  @WillAyd, I believe I have addressed your feedback. Please let me know what are the next steps. I see there are some failures in CI that seem to unrelated to the engine tests. Would be happy to take a closer look into these or work through them together.

MarcoGorelli reviewed (commented):
One question / consideration is whether pandas should just have a standardised entrypoint in `apply` such that third parties (like bodo and numba) can implement their own `apply` engine

Something like that if `engine=module_name` is specified, then `module_name` is imported and some standardised function is looked for, and that one is called from pandas. So, instead of
```python
        if self.engine == "numba":
            results, res_index = self.apply_series_numba()
        elif self.engine == "bodo":
            return self.apply_series_bodo()
        elif self.engine == "python":
            results, res_index = self.apply_series_generator()
```
to do
```python
        if self.engine == "python":
            results, res_index = self.apply_series_generator()
        else:
            module = importlib.import_module(self.engine)
            if not hasattr(module, '__pandas_apply_series_generator__'):
                raise ValueError(f"Custom engine {self.engine} is expected to implement '__pandas_apply_series_generator__'")
            results, res_index = module.__pandas_apply_series_generator__()
```

This could then
- lower the maintenance and testing burden on pandas
- open up the doors for more custom apply engines

The Numba engine already lives in pandas, I'm not suggesting to get rid of it. Rather, this could be an opportunity to "dogfood" the `'__pandas_apply_series_generator__'` - the Numba engine code within pandas lives in a Numba-wrapper which implements `'__pandas_apply_series_generator__'`. For any other engine, either the engine itself implements `'__pandas_apply_series_generator__'`, or users are free to make their own compliant wrappers

scott-routledge2 commented:
@MarcoGorelli Responded to your comment under the issue as well. We would be more than happy to explore this direction.

datapythonista committed:
Merge branch 'main' into scott/bodo_udf_engine

mroeschke commented on pandas/tests/util/test_bodo.py:
```diff
@@ -0,0 +1,22 @@
+
+def test_bodo_not_installed_df_apply():
+    "Test that importing bodo when not installed results in ImportError."
+    bodo_installed = bool(td.import_optional_dependency("bodo", errors="ignore"))
+    if bodo_installed:
+        pytest.skip("bodo is installed.")
```
```suggestion
    pytest.importorskip("bodo")
```

mroeschke commented on pandas/core/apply.py:
```diff
@@ -1089,6 +1105,36 @@ def apply_series_numba(self):
+
+        bodo = import_optional_dependency("bodo")
+
+        @bodo.jit(**self.engine_kwargs)
```
Does `boto.jit` take the same keyword arguments as `numba.jit`? I couldn't easily find the API docs for `bodo.jit`

mroeschke commented on pandas/core/apply.py:
```diff
@@ -1089,6 +1105,36 @@ def apply_series_numba(self):
+
+        try:
+            result = do_apply(self.obj, self.func, self.axis)
+        except bodo.utils.typing.BodoError as e:
+            raise ExecutionError("Execution with engine='bodo' failed.") from e
```
IMO I don't think pandas should need to define and raise a new exception class for BodoError. We don't do this for numba for example

WillAyd replied:
I think there is definitely value in offering up an abstracted error type as it makes writing generic code easier as an end user. Maybe this should be done as a pre-cursor with just Numba?

datapythonista replied:
Do you mind sharing a specific example of what you have in mind? I fail to understand yoir point, and it does seem like ir makes things more complex if users receive a pandas exception when Bodo can't run the user function. Unless the point is the error message, to let users know that it's not a problem in pandas ans rather it's in Bodo being able to execute the user function. That could make more sense if the errors the user receives from Bodo don't point the user in this direction.

WillAyd replied:
The analogy would be in SQL where we have many different engines and connections. Whether someone is using SQLAlchemy, ADBC, or SQLite3 we lean towards raising just `pd.DataBaseError` so that end users can write generic code like:

```
try:
    pd.read_sql(..., conn=<connection_type>)
except pd.DatabaseError as e:
    ...
```

If we required them to catch the specific error raised by the various connection types, they would have to adjust their error handling as well, which I think is less than ideal

mroeschke commented on pandas/tests/apply/test_bodo.py:
```diff
@@ -0,0 +1,145 @@
+pytestmark = [pytest.mark.single_cpu, pytest.mark.bodo_udf_engine]
+
+
+@pytest.fixture(params=["bodo"])
+def engine(request):
+    """Test bodo engine by itself to avoid extensions conflicting with numba.
+
+    Note: Using a fixture here to avoid importing at the start of the session.
+    """
+    if request.param == "bodo":
+        pytest.importorskip("bodo")
+    return request.param
```
```suggestion
@pytest.fixture
def engine(request):
    """Test bodo engine by itself to avoid extensions conflicting with numba.

    Note: Using a fixture here to avoid importing at the start of the session.
    """
    if request.param == "bodo":
        pytest.importorskip("bodo")
    return "bodo"
```

datapythonista commented:
I think this has been superseded by #61032. Let me know @scott-routledge2 if anyfeedback.
